// react-native-activity-feed-core/lib/styleguideComponents/PathlineRenderer.js
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.styles = undefined;
var _jsxFileName = 'src/styleguideComponents/PathlineRenderer.js';
exports.PathlineRenderer = PathlineRenderer;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Link = require('react-styleguidist/lib/rsg-components/Link');

var _Link2 = _interopRequireDefault(_Link);

var _Styled = require('react-styleguidist/lib/rsg-components/Styled');

var _Styled2 = _interopRequireDefault(_Styled);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styles = exports.styles = function styles(_ref) {
  var space = _ref.space,
      fontFamily = _ref.fontFamily,
      fontSize = _ref.fontSize,
      color = _ref.color;
  return {
    pathline: {
      fontFamily: fontFamily.monospace,
      fontSize: fontSize.small,
      color: color.light,
      wordBreak: 'break-all'
    },
    copyButton: {
      marginLeft: space[0]
    }
  };
};

function PathlineRenderer(_ref2) {
  var classes = _ref2.classes,
      children = _ref2.children;

  var source = children;

  return _react2.default.createElement(
    'div',
    { className: classes.pathline, __source: {
        fileName: _jsxFileName,
        lineNumber: 23
      }
    },
    _react2.default.createElement(
      _Link2.default,
      {
        href: 'https://github.com/GetStream/react-native-activity-feed/blob/master/' + source,
        target: 'blank',
        rel: 'noopener',
        __source: {
          fileName: _jsxFileName,
          lineNumber: 24
        }
      },
      source
    )
  );
}

PathlineRenderer.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  children: _propTypes2.default.string
};

exports.default = (0, _Styled2.default)(styles)(PathlineRenderer);
